File: src/main/java/at/hrechny/predictionsbot/mapper/CompetitionMapper.java
package at.hrechny.predictionsbot.mapper;

import at.hrechny.predictionsbot.config.MapperConfig;
import at.hrechny.predictionsbot.database.entity.CompetitionEntity;
import at.hrechny.predictionsbot.database.entity.SeasonEntity;
import at.hrechny.predictionsbot.model.Competition;
import org.apache.commons.collections4.CollectionUtils;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Named;

@Mapper(config = MapperConfig.class)
public interface CompetitionMapper {

  @Mapping(target = "active", source = "source", qualifiedByName = "isActive")
  Competition entityToModel(CompetitionEntity source);

  @Mapping(target = "seasons", ignore = true)
  CompetitionEntity modelToEntity(Competition source);

  @Named("isActive")
  default boolean isActive(CompetitionEntity entity) {
    return CollectionUtils.isNotEmpty(entity.getSeasons()) && entity.getSeasons().stream().anyMatch(SeasonEntity::isActive);
  }

}

File: src/main/kotlin/at/hrechny/predictionsbot/mapper/CompetitionMapper.kt
package at.hrechny.predictionsbot.mapper

import at.hrechny.predictionsbot.database.entity.CompetitionEntity
import at.hrechny.predictionsbot.database.entity.SeasonEntity
import at.hrechny.predictionsbot.controller.model.competition.CompetitionCreateRequestDto
import at.hrechny.predictionsbot.controller.model.competition.CompetitionResponseDto
import jakarta.inject.Singleton
import org.apache.commons.collections4.CollectionUtils

@Singleton
class CompetitionMapper {
    fun entityToModel(source: CompetitionEntity): CompetitionResponseDto =
        CompetitionResponseDto().apply {
            id = source.id
            name = source.name
            apiFootballId = source.apiFootballId
            active = isActive(source)
        }

    fun modelToEntity(source: CompetitionCreateRequestDto): CompetitionEntity =
        CompetitionEntity().apply {
            id = source.id
            name = source.name
            apiFootballId = source.apiFootballId
        }

    fun isActive(entity: CompetitionEntity): Boolean =
        CollectionUtils.isNotEmpty(entity.seasons) && entity.seasons.any(SeasonEntity::isActive)
}
